pg_search/src/postgres/customscan/pdbscan/exec_methods/fast_fields/mod.rs

```diff
@@ -405,3 +425,120 @@ pub fn estimate_cardinality(indexrel: &PgRelation, field: &str) -> Option<usize>
             .num_terms(),
     )
 }
+
+/// Process attributes using fast fields, creating a mapping and populating the datum array.
+/// This function is shared between the string and numeric fast field implementations.
+#[allow(clippy::too_many_arguments)]
+pub unsafe fn process_fast_fields(
+    natts: usize,
+    tupdesc: &PgTupleDesc<'_>,
+    which_fast_fields: &[WhichFastField],
+    fast_fields: &mut FFHelper,
+    slot: *mut pg_sys::TupleTableSlot,
+    scored: SearchIndexScore,
+    doc_address: DocAddress,
+    string_buffer: &mut Option<String>,
+) {
+    // Build attribute to fast field mapping
+    let mut attr_to_ff_map = std::collections::HashMap::new();
+
+    // Step 1: First try to match named attributes by name
+    for i in 0..natts {
+        if let Some(att) = tupdesc.get(i) {
+            let att_name = att.name().to_lowercase();
+            // Skip empty named attributes - handle them later
+            if !att_name.is_empty() {
+                // Try to find fast field with matching name
+                if let Some(idx) = which_fast_fields.iter().position(|ff| {
+                    if let WhichFastField::Named(name, _) = ff {
+                        name.to_lowercase() == att_name
+                    } else {
+                        false
+                    }
+                }) {
+                    attr_to_ff_map.insert(i, idx);
+                    continue;
+                }
+            }
+        }
+    }
+
+    // Step 2: Position-based matching for any remaining attributes
+    let mut next_ff_idx = 0;
+
+    // Simple position-based mapping, assuming attributes and fast fields are in the same order
+    for i in 0..natts {
+        if !attr_to_ff_map.contains_key(&i) {
+            // Find next unused fast field index
+            while next_ff_idx < which_fast_fields.len()
+                && attr_to_ff_map.values().any(|&v| v == next_ff_idx)
+            {
+                next_ff_idx += 1;
+            }
+
+            if next_ff_idx < which_fast_fields.len() {
+                attr_to_ff_map.insert(i, next_ff_idx);
+                next_ff_idx += 1;
+            }
+        }
+    }
+
+    // Ensure every attribute has a mapping
+    for i in 0..natts {
+        debug_assert!(
+            attr_to_ff_map.contains_key(&i),
+            "Attribute at position {} has no fast field mapping",
+            i
+        );
```

Comment:
These feel like they should be full asserts/`expects`...? Or do we not have the confidence for that yet?

Reply:
Actually a very good idea to panic! in cases that goes against our assumption. This should be our development style to avoid wrong output.